Mathlib/CategoryTheory/FiberedCategory/HomLift.lean

```diff
@@ -0,0 +1,210 @@
+/-
+Copyright (c) 2024 Calle Sönne. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Paul Lezeau, Calle Sönne
+-/
+
+import Mathlib.CategoryTheory.Functor.Category
+import Mathlib.CategoryTheory.CommSq
+
+/-!
+
+# HomLift
+
+Given a functor `p : 𝒳 ⥤ 𝒮`, this file provides API for expressing the fact that `p(φ) = f`
+for given morphisms `φ` and `f`.
+
+## Main definition
+
+Given morphism `φ : a ⟶ b` in `𝒳` and `f : R ⟶ S` in `𝒮`, `IsHomLift p f φ` is defined as a
+structure containing the data that `p(a) = R`, `p(b) = S` and the fact that the following square
+commutes
+```
+  p(a) --p(φ)--> p(b)
+  |               |
+  |               |
+  v               v
+  R -----f------> S
+```
+where the vertical arrows are given by `eqToHom` corresponding to the equalities `p(a) = R` and
+`p(b) = S`.
+
+-/
+
+universe u₁ v₁ u₂ v₂
+
+open CategoryTheory Functor Category
+
+variable {𝒮 : Type u₁} {𝒳 : Type u₂} [Category 𝒳] [Category 𝒮]
+
+/-- The proposition that an arrow a --φ--> b lifts an arrow R --f--> S in 𝒮 via p. This is
+often drawn as:
+```
+  a --φ--> b
+  -        -
+  |        |
+  v        v
+  R --f--> S
+``` -/
+structure IsHomLift (p : 𝒳 ⥤ 𝒮) {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ⟶ b) : Prop where
+  (ObjLiftDomain : p.obj a = R)
```

Comment:
```suggestion
  domain_eq : p.obj a = R
```
The syntax without the braces is prefered. Also, as this field is a "lemma", the initial should not be capitalized. The tactic `by aesop_cat` could be added as default so as to ease automation.

Did you consider making `IsHomLift` a type-class? It seems reasonable to think that it may be convenient. I would suggest the following:

```lean
class Functor.IsHomLift (p : 𝒳 ⥤ 𝒮) {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ⟶ b) : Prop where
  domain_eq : p.obj a = R := by aesop_cat
  codomain_eq : p.obj b = S := by aesop_cat
  fac : f = eqToHom domain_eq.symm ≫ p.map φ ≫ eqToHom codomain_eq := by aesop_cat

instance (p : 𝒳 ⥤ 𝒮) {a b : 𝒳} (f : a ⟶ b) : p.IsHomLift (p.map f) f where
instance (p : 𝒳 ⥤ 𝒮) (a : 𝒳) : p.IsHomLift (𝟙 (p.obj a)) (𝟙 a) where
```

Reply:
Thanks! I will experiment a bit with making it a class